rewrite of operators 101 (#405)
astronomer/airflow-guides: guides/what-is-an-operator.md

- If an operator exists for your specific use case, you should use it. This will make your DAGs easier to read and maintain, and is preferable to writing your own logic in a Python function or using [hooks](https://www.astronomer.io/guides/what-is-a-hook) directly.
- However, if an operator does not exist for your use case or you need to extend an existing operator, you may do so. For more on how to customize operators, check out our previous [Anatomy of an Operator webinar](https://www.astronomer.io/events/webinars/anatomy-of-an-operator).
- [Sensors](https://www.astronomer.io/guides/what-is-a-sensor) are a type of operator that wait for something to happen. They can be used to make your DAGs more event-driven.
- [Deferrable Operators](https://www.astronomer.io/guides/deferrable-operators) are a type of operator that release their worker slot while waiting for their work to be completed. This can result in cost savings and greater scalability, and Astronomer recommends using them whenever possible. Note that you must be using Airflow 2.2+ and have a triggerer running to use deferrable operators. 

> - Any operator that interacts with a service external to Airflow will typically require a connection so that Airflow can authenticate to that external system. More information on how to set up connections can be found our guide on [managing connections](https://www.astronomer.io/guides/connections/) or in the example section below.

Reviewer: ```suggestion
- Any operator that interacts with a service external to Airflow will typically require a connection so that Airflow can authenticate to that external system. More information on how to set up connections can be found in our guide on [managing connections](https://www.astronomer.io/guides/connections/) or in the example section below.
```

## Example Implementation

This example shows how to use several common operators in a DAG used to transfer data from S3 to Redshift and perform data quality checks. 

> **Note:** The full code and repository for this example can be found on the [Astronomer Registry](https://registry.astronomer.io/dags/simple-redshift-3).

The following operators are used:

- [EmptyOperator](https://registry.astronomer.io/providers/apache-airflow/modules/dummyoperator): This operator is part of core Airflow and does nothing. It is used to organize the flow of tasks in the DAG.

> - [PythonDecoratedOperator](https://registry.astronomer.io/providers/apache-airflow/modules/pythonoperator): This operator is part of core Airflow and executes a Python function. It is functionally the same as the PythonOperator, but it instantiated using the `@task` decorator.

Reviewer: ```suggestion
- [PythonDecoratedOperator](https://registry.astronomer.io/providers/apache-airflow/modules/pythonoperator): This operator is part of core Airflow and executes a Python function. It is functionally the same as the PythonOperator, but it instantiated using the `@task` [decorator](https://www.astronomer.io/guides/airflow-decorators/).
```

- [PostgresOperator](https://registry.astronomer.io/providers/postgres/modules/postgresoperator): This operator is part of the [Postgres provider](https://registry.astronomer.io/providers/postgres) and is used to execute a query against a Postgres database.
- [SQLCheckOperator](https://registry.astronomer.io/providers/apache-airflow/modules/sqlcheckoperator): This operator is part of core Airflow and is used to perform checks against a database using a SQL query.

The following code shows how each of those operators can be instantiated in a DAG file to define the pipeline:

```python
import hashlib
import json

from airflow import DAG, AirflowException
from airflow.decorators import task
from airflow.models import Variable
from airflow.models.baseoperator import chain
from airflow.operators.empty_operator import EmptyOperator
from airflow.utils.dates import datetime
from airflow.providers.amazon.aws.hooks.s3 import S3Hook
from airflow.providers.amazon.aws.transfers.local_to_s3 import (

>     LocalFilesystemToS3Operator,

Reviewer: ```suggestion
    LocalFilesystemToS3Operator
```

)

> from airflow.providers.amazon.aws.transfers.s3_to_redshift import S3ToRedshiftOperator

Reviewer: ```suggestion
from airflow.providers.amazon.aws.transfers.s3_to_redshift import (
    S3ToRedshiftOperator
)
```

from airflow.providers.postgres.operators.postgres import PostgresOperator
from airflow.operators.sql import SQLCheckOperator
from airflow.utils.task_group import TaskGroup

> # The file(s) to upload shouldn't be hardcoded in a production setting, this is just for demo purposes.

Reviewer: ```suggestion
# The file(s) to upload shouldn't be hardcoded in a production setting, 
# this is just for demo purposes.
```

CSV_FILE_NAME = "forestfires.csv"
CSV_FILE_PATH = f"include/sample_data/forestfire_data/{CSV_FILE_NAME}"

with DAG(
    "simple_redshift_3",
    start_date=datetime(2021, 7, 7),

>     description="A sample Airflow DAG to load data from csv files to S3 and then Redshift, with data integrity and quality checks.",

Reviewer: ```suggestion
    description="""A sample Airflow DAG to load data from csv files to S3 
                 and then Redshift, with data integrity and quality checks.""",
```

    schedule_interval=None,
    template_searchpath="/usr/local/airflow/include/sql/redshift_examples/",
    catchup=False,
) as dag:

    """

>     Before running the DAG, set the following in an Airflow or Environment Variable:

Reviewer: ```suggestion
    Before running the DAG, set the following in an Airflow 
    or Environment Variable:
```

    - key: aws_configs

>     - value: { "s3_bucket": [bucket_name], "s3_key_prefix": [key_prefix], "redshift_table": [table_name]}

Reviewer: ```suggestion
    - value: { "s3_bucket": [bucket_name], "s3_key_prefix": [key_prefix],
             "redshift_table": [table_name]}
```

    Fully replacing [bucket_name], [key_prefix], and [table_name].
    """

    upload_file = LocalFilesystemToS3Operator(
        task_id="upload_to_s3",
        filename=CSV_FILE_PATH,
        dest_key="{{ var.json.aws_configs.s3_key_prefix }}/" + CSV_FILE_PATH,
        dest_bucket="{{ var.json.aws_configs.s3_bucket }}",
        aws_conn_id="aws_default",
        replace=True,
    )

    @task
    def validate_etag():
        """
        #### Validation task

>         Check the destination ETag against the local MD5 hash to ensure the file

Reviewer: ```suggestion
        Check the destination ETag against the local MD5 hash to ensure 
        the file was uploaded without errors.
```

>         was uploaded without errors.

Reviewer: ```suggestion
```

        """
        s3 = S3Hook()
        aws_configs = Variable.get("aws_configs", deserialize_json=True)
        obj = s3.get_key(
            key=f"{aws_configs.get('s3_key_prefix')}/{CSV_FILE_PATH}",
            bucket_name=aws_configs.get("s3_bucket"),
        )
        obj_etag = obj.e_tag.strip('"')
        # Change `CSV_FILE_PATH` to `CSV_CORRUPT_FILE_PATH` for the "sad path".
        file_hash = hashlib.md5(
            open(CSV_FILE_PATH).read().encode("utf-8")).hexdigest()
        if obj_etag != file_hash:
            raise AirflowException(

>                 f"Upload Error: Object ETag in S3 did not match hash of local file."

Reviewer: ```suggestion
                f"""Upload Error: Object ETag in S3 did not match 
                hash of local file."""
```

            )

>     validate_file = validate_etag()

Reviewer: ```suggestion
    # Tasks that were created using decorators have to be called to be used
    validate_file = validate_etag()
```

Adding this because at the very beginning I once spent way too much time staring at my decorated tasks not showing up in my dag because I did not know about this. 😅

Author: Good call!

    #### Create Redshift Table
    create_redshift_table = PostgresOperator(
        task_id="create_table",
        sql="create_redshift_forestfire_table.sql",
        postgres_conn_id="redshift_default",
    )

>     #### Second load task

Reviewer: There is an extra empty line before this line I don't know how to delete that in a suggestion, sry 😅

Author: I think you can do it if you do a multi-line suggestion. But I fixed it :)

        params={"filename": CSV_FILE_NAME},
    )

    #### Row-level data quality check
    with open("include/validation/forestfire_validation.json") as ffv:
        with TaskGroup(group_id="row_quality_checks") as quality_check_group:
            ffv_json = json.load(ffv)
            for id, values in ffv_json.items():
                values["id"] = id
                SQLCheckOperator(
                    task_id=f"forestfire_row_quality_check_{id}",
                    conn_id="redshift_default",
                    sql="row_quality_redshift_forestfire_check.sql",
                    params=values,
                )

    #### Drop Redshift table
    drop_redshift_table = PostgresOperator(
        task_id="drop_table",
        sql="drop_redshift_forestfire_table.sql",
        postgres_conn_id="redshift_default",
    )

    begin = EmptyOperator(task_id="begin")
    end = EmptyOperator(task_id="end")

    #### Define task dependencies
    chain(
        begin,
        upload_file,
        validate_file,
        create_redshift_table,
        load_to_redshift,
        validate_redshift,
        quality_check_group,
        drop_redshift_table,

>         end,

Reviewer: ```suggestion
        end
```

- The core Airflow package contains basic operators such as the PythonOperator and BashOperator, which will be automatically available in your Airflow environment. All other operators are part of provider packages, which must be installed separately. For example, the SnowflakeOperator is part of the [Snowflake provider](https://registry.astronomer.io/providers/snowflake).
- If an operator exists for your specific use case, you should use it. This will make your DAGs easier to read and maintain, and is preferable to writing your own logic in a Python function or using [hooks](https://www.astronomer.io/guides/what-is-a-hook) directly.
- However, if an operator does not exist for your use case or you need to extend an existing operator, you may do so. For more on how to customize operators, check out our previous [Anatomy of an Operator webinar](https://www.astronomer.io/events/webinars/anatomy-of-an-operator).
- [Sensors](https://www.astronomer.io/guides/what-is-a-sensor) are a type of operator that wait for something to happen. They can be used to make your DAGs more event-driven.

> - [Deferrable Operators](https://www.astronomer.io/guides/deferrable-operators) are a type of operator that release their worker slot while waiting for their work to be completed. This can result in cost savings and greater scalability, and Astronomer recommends using them whenever possible. Note that you must be using Airflow 2.2+ and have a triggerer running to use deferrable operators. 

Author: Yes, worth it. Updated the sentence a bit to add some of that nuance.

> - Any operator that interacts with a service external to Airflow will typically require a connection so that Airflow can authenticate to that external system. More information on how to set up connections can be found in our guide on [managing connections](https://www.astronomer.io/guides/connections/) or in the example section below.

Reviewer 2: ```suggestion
- Any operator that interacts with a service external to Airflow will typically require a connection so that Airflow can authenticate to that external system. More information on how to set up connections can be found in our guide on [managing connections](https://www.astronomer.io/guides/connections/) or in the examples to follow.
```
